### ISSUE TITLE
TORCH_CUDA_ARCH_LIST format

### ISSUE THREAD

**Issue by ydaiming** [pytorch/audio#1940]
### 🐛 Describe the bug

At the [pytorch/Dockerfile#L48](https://github.com/pytorch/pytorch/blob/master/Dockerfile#L48), `TORCH_CUDA_ARCH_LIST` are separated by spaces, while inside the builder of `torchaudio`, I found that the parser splits the list based on `;`: https://github.com/pytorch/audio/blob/e3734feff3df2fd4c81572ba84ca7b8d52c4fab2/tools/setup_helpers/extension.py#L104

It took me a long time to locate the source of errors during build time for multiple CUDA archs from source. Could you standardize the format for `TORCH_CUDA_ARCH_LIST`? Either by making all of them separated by `;`, or let the audio builder also able to parse spaces as splitters? 

### Versions

```
root@fe1fec0b5ec4:/# python collect_env.py
Collecting environment information...
PyTorch version: N/A
Is debug build: N/A
CUDA used to build PyTorch: N/A
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.3 LTS (x86_64)
GCC version: (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0
Clang version: Could not collect
CMake version: version 3.16.3
Libc version: glibc-2.31

Python version: 3.8.10 | packaged by conda-forge | (default, May 11 2021, 07:01:05)  [GCC 9.3.0] (64-bit runtime)
Python platform: Linux-5.4.0-1058-aws-x86_64-with-glibc2.10
Is CUDA available: N/A
CUDA runtime version: 11.3.109
GPU models and configuration: 
GPU 0: Tesla V100-SXM2-16GB
GPU 1: Tesla V100-SXM2-16GB
GPU 2: Tesla V100-SXM2-16GB
GPU 3: Tesla V100-SXM2-16GB
GPU 4: Tesla V100-SXM2-16GB
GPU 5: Tesla V100-SXM2-16GB
GPU 6: Tesla V100-SXM2-16GB
GPU 7: Tesla V100-SXM2-16GB

Nvidia driver version: 450.142.00
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.2.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.2.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.2.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.2.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.2.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.2.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.2.0
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.20.3
[pip3] torch-model-archiver==0.4.2
[pip3] torchaudio==0.10.0+cu113
[pip3] torchserve==0.4.2
[pip3] torchvision==0.11.1+cu113
[conda] blas                      1.0                         mkl  
[conda] libblas                   3.9.0            11_linux64_mkl    conda-forge
[conda] libcblas                  3.9.0            11_linux64_mkl    conda-forge
[conda] liblapack                 3.9.0            11_linux64_mkl    conda-forge
[conda] liblapacke                3.9.0            11_linux64_mkl    conda-forge
[conda] magma-cuda113             2.5.2                         1    pytorch
[conda] mkl                       2021.3.0           h06a4308_520  
[conda] mkl-include               2021.3.0           h06a4308_520  
[conda] mkl-service               2.4.0            py38h7f8727e_0  
[conda] mkl_fft                   1.3.1            py38hd3c417c_0  
[conda] mkl_random                1.2.2            py38h51133e4_0  
[conda] numpy                     1.20.3                   pypi_0    pypi
[conda] torch-model-archiver      0.4.2                    pypi_0    pypi
[conda] torchaudio                0.10.0+cu113             pypi_0    pypi
[conda] torchserve                0.4.2                    pypi_0    pypi
[conda] torchvision               0.11.1+cu113             pypi_0    pypi
```

**Comment by ydaiming** [comment]
@malfet is there an update on this topic?

**Comment by malfet** [comment]
Hmm, standardizing sounds like a good idea as spaces as separators are specified in 
https://github.com/pytorch/pytorch/blob/114ef8c5ea291ae7744c33fa7c0d23b67e75d525/torch/utils/cpp_extension.py#L896
which is part of [C++ extension](https://pytorch.org/docs/stable/cpp_extension.html) documentation
The only thing that slightly confuses me, is that fact that same semicolon separated list is used in torchvision. 
@ydaiming have you file an issue against torchvision as well by any chance?

**Comment by fmassa** [comment]
Note that if we drop support for semi-colon, a lot of downstream projects will have to be changed as well, for example detectron2 https://github.com/facebookresearch/detectron2/blob/aa8ea943411a2e5cd616e4d517215779a2ea2dad/dev/packaging/pkg_helpers.bash#L23

Also, looks like most occurrences on GitHub uses semi-colon https://github.com/search?q=TORCH_CUDA_ARCH_LIST&type=code

**Comment by mthrok** [comment]
@fmassa This is not about dropping the support, but about how to define the build customization of domain libraries. I checked the detctron2's source and it does not seem to depend on torchvision at C++ level at build time, so it won't be affected.

The semicolon version happens to work because of the logic here.

https://github.com/pytorch/pytorch/blob/6011c35a7934a1528c2db02c40752aa3822ccfa6/torch/utils/cpp_extension.py#L1609-L1613
So using semicolon just for building a downstream project is fine. But if such customization is re-exposed to further downstream users, it should behave same as PyTorch core.

**Comment by mthrok** [comment]
Addressed via #2050
